### PR TITLE
Implement additional text object commands

### LINKED ISSUE #449: Implement i< and i(, and i{ text objects
I am going to send a PR for these shortly.

### PR DESCRIPTION
**Pull request by sectioneight**
This allows for e.g. 'ci(' and 'da<' commands.

Fixes #449

### CODE REVIEW

**Comment by sectioneight** [comment]
To be clear, any and all feedback, no matter how minor or how major, is appreciated. I'd like to continue to contribute to the project, and I'd like to write idiomatic TypeScript, but I'm nearly certain this code is not optimal as-is.

File: src/actions/actions.ts [hunk 1]
```diff
@@ -2190,11 +2188,11 @@ class MoveToMatchingBracket extends BaseMovement {
       let matchedPosition: Position | undefined = undefined;
 
       for (const { char, pos } of Position.IterateDocument(position, toFind.nextMatchIsForward)) {
-        if (char === charToMatch) {
+        if (char === charToMatch && charToMatch !== toFind.match) {
           stackHeight++;
         }
 
-        if (char === this.pairings[charToMatch].match) {
+        if (char === toFind.match) {
```
Comment (sectioneight) [inline src/actions/actions.ts]: All of the callers of this function (`this.nextBracket`) are already passing an object from `this.pairings[char]` as the `toFind`, so I think this is more correct. This also allows callers of this function to use characters that don't exist in `this.pairings`, such as `"` and `'`

File: src/actions/actions.ts [hunk 2]
```diff
@@ -2233,7 +2236,7 @@ class MoveToMatchingBracket extends BaseMovement {
     }
 
     return this.nextBracket(position, charToMatch, toFind, true);
-}
+  }
```
Comment (sectioneight) [inline src/actions/actions.ts]: Unrelated style cleanup, not sure if this belongs in this PR or not.

File: src/actions/actions.ts [hunk 3]
```diff
@@ -2246,6 +2249,48 @@ class MoveToMatchingBracket extends BaseMovement {
...
+  modes = [ModeName.Normal, ModeName.Visual, ModeName.VisualLine];
+  protected charToMatch: string;
+
+  public async execAction(position: Position, vimState: VimState): Promise<Position | IMovement> {
+    const text = TextEditor.getLineAt(position).text;
+    const pairedCharacter = this.pairings[this.charToMatch]
+    const charToClose = pairedCharacter ? pairedCharacter.match : this.charToMatch;
+
+    // First, search backwards for the opening character of the sequence
+    let start = -1;
```
Comment (sectioneight) [inline src/actions/actions.ts]: Is there a more idiomatic way to accomplish this without looping backwards? It feels weird using these -1 sentinel values, and using a plain for loop, but String.indexOf can only search forwards.

File: src/actions/actions.ts [hunk 4]
```diff
@@ -2246,6 +2249,48 @@ class MoveToMatchingBracket extends BaseMovement {
...
+      }
+    }
+    const startPos = new Position(position.line, start + 1);
+    const endPos = this.nextBracket(startPos, this.charToMatch, { match: charToClose, nextMatchIsForward: true }, false);
+    if (start == -1 || endPos.character == start + 1 || start > position.character || endPos.character < position.character ) {
+      return position;
+    }
+
+    return {
+      start : startPos,
```
Comment (sectioneight) [inline src/actions/actions.ts]: Not sure about style here. Was following some surrounding style, but VSCode seems to want to format this as `start: startPos` (no space before opening colon).

File: src/actions/actions.ts [hunk 5]
```diff
@@ -2246,6 +2249,48 @@ class MoveToMatchingBracket extends BaseMovement {
...
+
+@RegisterAction
+class MoveIParentheses extends MoveInsideCharacter {
+  keys = ["i", "("];
+  charToMatch = "(";
+}
+
+@RegisterAction
+class MoveIDoubleQuote extends MoveInsideCharacter {
+  keys = ["i", "\""];
```
Comment (sectioneight) [inline src/actions/actions.ts]: This feels kind of silly. Is there a way to set the `keys` property programatically and just register a single action/class that can handle all of these? (since there are more characters to implement, but wanted to start small)

File: test/mode/modeNormal.test.ts [hunk 6]
```diff
@@ -220,6 +220,27 @@ suite("Mode Normal", () => {
     });
 
     newTest({
+      title: "Can handle 'ci(' on first parentheses",
```
Comment (sectioneight) [inline test/mode/modeNormal.test.ts]: Is there a way to run a single test case?

File: src/actions/actions.ts [hunk 7]
```diff
@@ -2246,6 +2249,48 @@ class MoveToMatchingBracket extends BaseMovement {
...
+    // First, search backwards for the opening character of the sequence
+    let start = -1;
+    for (let i = position.character; i >= 0; i--) {
+      if (text[i] == this.charToMatch) {
+        start = i;
+        break;
+      }
+    }
+    const startPos = new Position(position.line, start + 1);
+    const endPos = this.nextBracket(startPos, this.charToMatch, { match: charToClose, nextMatchIsForward: true }, false);
```
Comment (sectioneight) [inline src/actions/actions.ts]: This is where the behavior differs from vim. This should be calling some variant of `nextBracket` that only iterates positions up to the end of the current line, not the whole document.

File: src/actions/actions.ts (hunk 2 above)
Comment (johnfn) [inline src/actions/actions.ts]: No worries; I'm totally fine with you sneaking in stuff like this.

File: src/actions/actions.ts (hunk 5 above)
Comment (johnfn) [inline src/actions/actions.ts]: I am actually okay with this. I'd prefer to explicitly list out every key combination that exists, even though it's slightly less DRY than we'd otherwise want.

File: test/mode/modeNormal.test.ts (hunk 6 above)
Comment (jpoon) [inline test/mode/modeNormal.test.ts]: I haven't found a way :(

File: src/actions/actions.ts (hunk 1 above)
Comment (johnfn) [inline src/actions/actions.ts]: Good catch! That must have been vestigial.

I would also suggest adding ' and " back to our `pairings` list, perhaps adding a flag so that we don't match them with %.

File: test/mode/modeNormal.test.ts (hunk 6 above)
Comment (johnfn) [inline test/mode/modeNormal.test.ts]: There is a way!

newTestOnly

File: src/actions/actions.ts (hunk 4 above)
Comment (johnfn) [inline src/actions/actions.ts]: No hard rules; I just like to line up colons sometimes.

**Comment by johnfn** [comment]
This is looking good so far, and I'm pretty hyped to have this functionality! 😀 

The only thing I'd suggest to change would be how you use inheritance to share the matching functionality with `MoveToMatchingBase`. You could just extract the function out into a new file and our class hierarchy would become 1 less deep, which would be nice.

File: src/actions/actions.ts (hunk 1 above)
Comment (rebornix) [inline src/actions/actions.ts]: This `pairings` bothers me a lot when I implement [MoveToUnclosedRoundBracketBackward](https://github.com/sectioneight/Vim/blob/5a50f2105119bd8cb7ba4a105560a76b2369bb59/src/actions/actions.ts#L2294). And the first hack I did is moving the bracket matching to a separate method called `nextBracket` in https://github.com/vscodevim/Vim/commit/b0a203830c7cda40100c678bd79c2e45c495af94 but now it seems `nextBracket` is no longer a good name as we might match quotes. Something called `nextPairedChar` looks better to me.

Besides, your PR is good enough for this fix but you may want to take a look at [all text objects commands](https://github.com/VSCodeVim/Vim/blob/master/ROADMAP.md#text-objects-only-in-visual-mode-or-after-an-operator), maybe we can make this `abstract class MoveToMatchingBase` more generic?

File: src/actions/actions.ts (hunk 5 above)
Comment (sectioneight) [inline src/actions/actions.ts]: Sounds good to me. Just didn't want to leave you with some massive copy-pasta mess if there was a better way.

File: src/actions/actions.ts (hunk 1 above)
Comment (sectioneight) [inline src/actions/actions.ts]: Yes, I plan on making the class generic enough so that it can handle `a<`, `i)` etc (and may add them to this PR now that I have some better confidence I'm going down the right track).

Will come up with some better naming, and add `'`/`"` into `pairings`

**Comment by sectioneight** [comment]
Ok, I think I have most of what I need to get this PR to a merge-able state.

One open question that I left at the top: is there a way to return something to indicate that this **didn't** match, so that for example an unsuccessful `ci"` won't switch to insert mode?

**Comment by sectioneight** [comment]
Needs a few more tests, but I think this is almost there.

**Comment by johnfn** [comment]
Ahh. Your question about a failed ci" is actually something I haven't quite got around to yet. For now could you just mark it with a TODO and return a range of length 0? It's not perfect, I admit.

File: src/actions/actions.ts [hunk 8]
```diff
@@ -2154,68 +2155,19 @@ class MoveToMatchingBracket extends BaseMovement {
...
-      // TODO(bell)
-      return position;
-  }
+  matcher = new PairMatcher();
 
   public async execAction(position: Position, vimState: VimState): Promise<Position> {
     const text = TextEditor.getLineAt(position).text;
     const charToMatch = text[position.character];
-    const toFind = this.pairings[charToMatch];
+    const toFind = PairMatcher.pairings[charToMatch];
```
Comment (johnfn) [inline src/actions/actions.ts]: What do you think about making these static methods rather than instance methods? There isn't really any associated state we need to carry along, is there?

File: src/matching/matcher.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+export class PairMatcher {
+  static pairings: { [key: string]: { match: string, nextMatchIsForward: boolean, limited?: boolean }} = {
+    "(" : { match: ")",  nextMatchIsForward: true  },
+    "{" : { match: "}",  nextMatchIsForward: true  },
+    "[" : { match: "]",  nextMatchIsForward: true  },
+    ")" : { match: "(",  nextMatchIsForward: false },
+    "}" : { match: "{",  nextMatchIsForward: false },
+    "]" : { match: "[",  nextMatchIsForward: false },
+    // These characters can't be used for "%"-based matching, but are still
+    // useful for text objects. TODO limited is a terrible name.
```
Comment (johnfn) [inline src/matching/matcher.ts]: Haha, nice, I was going to suggest you change limited to something else also. 

matchesWithPercentageMotion might be ok.

File: src/actions/actions.ts (hunk 8 above)
Comment (sectioneight) [inline src/actions/actions.ts]: Yea, I originally had it being stateful (where you initialized a PairMatcher with the character), but that seemed wasteful as well. Wasn't sure about preferences for static/non-static. Will change to static.

File: src/matching/matcher.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    "'" : { match: "'",  nextMatchIsForward: true, limited: true },
+    "\"": { match: "\"", nextMatchIsForward: true, limited: true },
+    "<" : { match: ">",  nextMatchIsForward: true, limited: true },
+  };
+
+  previousChar(position: Position, charToMatch: string, currentLineOnly: boolean = false): Position {
```
Comment (johnfn) [inline src/matching/matcher.ts]: This strikes me as something that belongs on Position. Maybe cal it lastIndexOf to indicate similarity with the string method of the same name.

File: src/actions/actions.ts (hunk 8 above)
Comment (johnfn) [inline src/actions/actions.ts]: Yeah. I prefer static to indicate stateless, though unfortunately there are places where I don't adhere to that.

File: src/matching/matcher.ts (hunk 10 above)
Comment (sectioneight) [inline src/matching/matcher.ts]: Indeed. Done.

**Comment by sectioneight** [comment]
By "return a range of length 0" do you mean an IMovement with the same start/stop? I tried that, but it still didn't "cancel" the insert/change.

**Comment by johnfn** [comment]
Yeah, there's no way to cancel, so we just return an empty selection. It's not great- we'll have to fix it later.

**Comment by sectioneight** [comment]
Yea, I understand that there isn't currently a cancel. Maybe there's a bug in my code, but when I return a `{start: pos, stop: pos}` (which I think is what you meant), a `ci"` still deletes the first character.

**Comment by rebornix** [comment]
@sectioneight these commands are on my list before I worked on Window/Tab and CJK stuff. Good to see you can help on this, they are quite important.

**Comment by sectioneight** [comment]
Disregard my previous comment, `ci"` without a match is not deleting any characters (only changing to insert mode, which is sub-par but not a dealbreaker)

**Comment by sectioneight** [comment]
NOTE: There is one limitation to this that I just discovered.

If you have:

```
"foo"
```

and you have your cursor on the closing `"` mark, and type e.g. `ci"`, it won't recognize the match while vim will.

I don't consider it a blocker, but if there's an obvious way to fix this I'd be open to doing it now.

**Comment by johnfn** [comment]
Ahh, yes. When I initially said that i" and i' would be harder, that was a bit of a concern of mine. I'm not convinced there's an easier way than parsing the entire line.

**Comment by rebornix** [comment]
We are using `inclusive` parameter quite often in [Position](https://github.com/VSCodeVim/Vim/blob/master/src/motion/position.ts#L184), I suppose that's also what we need for this generic function.

I'd like to see this PR merged as I don't want to see it Open for too long. But just make sure we don't release this feature without a fix, people are likely to run `i"` while the cursor is on a `"`.

**Comment by sectioneight** [comment]
To be clear, the limitation is only on the closing quote. ci" on an opening quote behaves as expected, as does `ci(` on the opening **or** closing parens, as well as inside. Because of the observed behavior, I'm pretty certain it's due to `nextPairedChar` not correctly handling the closing and opening chars being the same.

**Comment by sectioneight** [comment]
My current thinking: handling quotes is going to take some more work, and I want to balance making forward progress with not introducing buggy code. I'm going to remove the single- and double-quote functionality from this PR and work out the proper algorithm to handle them (since it's really a different problem than handling the rest).

That way we can (hopefully) get this PR merged this weekend and I'll spend some more time working through the edge cases of properly handling quotes.

**Comment by sectioneight** [comment]
Ok, PR updated, I think the functionality contained is complete (although I marked it as "partially supported" on the roadmap until at least one other person tries it out and says it works).

**Comment by johnfn** [comment]
Awesome work! This looks good to me.
